File: src/main/java/com/softwareloop/glo/DatStore.java
package com.softwareloop.glo;

import com.softwareloop.glo.model.*;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.FilenameUtils;
import org.xml.sax.InputSource;
import org.xml.sax.XMLReader;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.Unmarshaller;
import javax.xml.parsers.SAXParserFactory;
import javax.xml.transform.sax.SAXSource;
import java.io.IOException;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Slf4j
public class DatStore {

    //--------------------------------------------------------------------------
    // Constants
    //--------------------------------------------------------------------------

    public static final Pattern MD5_PATTERN = Pattern.compile("[A-F0-9]{32}");

    //--------------------------------------------------------------------------
    // Fields
    //--------------------------------------------------------------------------

    private final JAXBContext jaxbContext;
    private final SAXParserFactory spf;
    private final Map<String, List<RomSummary>> md5Map;

    //--------------------------------------------------------------------------
    // Constructors
    //--------------------------------------------------------------------------

    @SneakyThrows
    public DatStore() {
        jaxbContext = JAXBContext.newInstance(Datafile.class);
        spf = SAXParserFactory.newInstance();
        spf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
        spf.setFeature("http://xml.org/sax/features/validation", false);
        md5Map = new HashMap<>();
    }


    //--------------------------------------------------------------------------
    // Public methods
    //--------------------------------------------------------------------------

    public void loadDatDir(Path datDir) throws IOException {
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(datDir)) {
            for (Path datFile : stream) {
                log.debug("Reading: {}", datFile);
                String fileName = datFile.getFileName().toString();
                String extension = FilenameUtils.getExtension(fileName);
                if (Files.isRegularFile(datFile) && "dat".equalsIgnoreCase(extension)) {
                    Datafile datafile = loadDatFile(datFile);
                    add(datafile);
                }
            }
        }
    }

    @SneakyThrows
    public Datafile loadDatFile(Path datFile) {
        try (Reader reader = Files.newBufferedReader(datFile, StandardCharsets.UTF_8)) {
            XMLReader xmlReader = spf.newSAXParser().getXMLReader();
            InputSource inputSource = new InputSource(reader);
            SAXSource source = new SAXSource(xmlReader, inputSource);
            Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
            return (Datafile) jaxbUnmarshaller.unmarshal(source);
        }
    }

    public void add(Datafile datafile) {
        Header header = datafile.getHeader();
        String datName = header.getName();
        for (Game game : datafile.getGames()) {
            for (Rom rom : game.getRoms()) {
                String romMd5 = rom.getMd5();
                if (romMd5 == null) {
                    log.trace("md5 is null, skipping");
                    continue;
                }
                romMd5 = romMd5.toUpperCase();
                Matcher md5Matcher = MD5_PATTERN.matcher(romMd5);
                if (!md5Matcher.matches()) {
                    log.trace("md5 has invalid format, skipping");
                    continue;
                }
                RomSummary romSummary = new RomSummary();
                romSummary.setDatName(datName);
                romSummary.setRomName(rom.getName());
                romSummary.setRomSize(rom.getSize());
                romSummary.setRomMd5(romMd5);
                List<RomSummary> romSummaries =
                        md5Map.computeIfAbsent(romMd5, k -> new ArrayList<>());
                if (!romSummaries.contains(romSummary)) {
                    romSummaries.add(romSummary);
                }
            }
        }
    }

    public List<RomSummary> getRomSummaryByMd5(String md5) {
        return md5Map.get(md5);
    }


    //--------------------------------------------------------------------------
    // Methods
    //--------------------------------------------------------------------------

}

File: src/main/java/com/softwareloop/glo/DatStore.kt
package com.softwareloop.glo

import com.softwareloop.glo.model.Datafile
import com.softwareloop.glo.model.RomSummary
import org.apache.commons.io.FilenameUtils
import org.xml.sax.InputSource
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import java.util.regex.Pattern
import javax.xml.bind.JAXBContext
import javax.xml.parsers.SAXParserFactory
import javax.xml.transform.sax.SAXSource

class DatStore {

    //--------------------------------------------------------------------------
    // Constants
    //--------------------------------------------------------------------------

    companion object {
        val MD5_PATTERN: Pattern = Pattern.compile("[A-F0-9]{32}")
        val ROM_NAME_ILLEGAL_PATTERN: Pattern = Pattern.compile("[/\\\\]")
        val jaxbContext: JAXBContext = JAXBContext.newInstance(Datafile::class.java)
        val spf: SAXParserFactory = SAXParserFactory.newInstance().apply {
            setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)
            setFeature("http://xml.org/sax/features/validation", false)
        }
    }

    //--------------------------------------------------------------------------
    // Fields
    //--------------------------------------------------------------------------

    val md5Map = HashMap<String, MutableList<RomSummary>>()

    //--------------------------------------------------------------------------
    // Public methods
    //--------------------------------------------------------------------------

    fun loadDatDir(datDir: Path) {
        Files.newDirectoryStream(datDir).use { stream ->
            var nDats = 0
            for (datFile in stream) {
                Log.debug("Reading: %s", datFile)
                val fileName = datFile.fileName.toString()
                val extension = FilenameUtils.getExtension(fileName)
                if (Files.isRegularFile(datFile) && "dat".equals(extension, ignoreCase = true)) {
                    val datafile = loadDatFile(datFile)
                    add(datafile)
                    nDats++
                }
            }
            Log.info("Loaded %d dat files and %d entries", nDats, md5Map.size)
        }
    }

    fun loadDatFile(datFile: Path): Datafile {
        Files.newBufferedReader(datFile, StandardCharsets.UTF_8).use { reader ->
            val xmlReader = spf.newSAXParser().xmlReader
            val inputSource = InputSource(reader)
            val source = SAXSource(xmlReader, inputSource)
            val jaxbUnmarshaller = jaxbContext.createUnmarshaller()
            return jaxbUnmarshaller.unmarshal(source) as Datafile
        }
    }

    fun add(datafile: Datafile) {
        val header = datafile.header ?: return
        val datName = header.name ?: return
        val games = datafile.games ?: return
        for (game in games) {
            val roms = game.roms ?: return
            for (rom in roms) {
                val romName = rom.name ?: continue
                val romNameMatcher = ROM_NAME_ILLEGAL_PATTERN.matcher(romName)
                if (romNameMatcher.find()) {
                    Log.debug("Rom name '%s'contains illegal characters, skipping.", romName)
                }
                var romMd5 = rom.md5
                if (romMd5 == null) {
                    Log.debug("md5 is null, skipping")
                    continue
                }
                romMd5 = romMd5.uppercase(Locale.getDefault())
                val md5Matcher = MD5_PATTERN.matcher(romMd5)
                if (!md5Matcher.matches()) {
                    Log.debug("md5 has invalid format, skipping")
                    continue
                }
                val romSummary = RomSummary(datName, romName)
                val romSummaries = md5Map.computeIfAbsent(romMd5) { k: String? -> ArrayList() }
                if (!romSummaries.contains(romSummary)) {
                    romSummaries.add(romSummary)
                }
            }
        }
    }

    fun getRomSummaryByMd5(md5: String): List<RomSummary>? {
        return md5Map[md5]
    }

}